lib/SPIRVProducerPass.cpp

```diff
@@ -1925,716 +1900,598 @@ spv::BuiltIn SPIRVProducerPass::GetBuiltin(StringRef Name) const {
-      auto *Inst = new SPIRVInstruction(spv::OpTypeVoid, nextID++);
-      SPIRVInstList.push_back(Inst);
-      break;
+      Ops << MkId(getSPIRVType(ParamTy));
     }
-    case Type::FunctionTyID: {
-      // Generate SPIRV instruction for function type.
-      FunctionType *FTy = cast<FunctionType>(Ty);
-
-      // Ops[0] = Return Type ID
-      // Ops[1] ... Ops[n] = Parameter Type IDs
-      SPIRVOperandList Ops;
-
-      // Find SPIRV instruction for return type
-      Ops << MkId(lookupType(FTy->getReturnType()));
-
-      // Find SPIRV instructions for parameter types
-      for (unsigned k = 0; k < FTy->getNumParams(); k++) {
-        // Find SPIRV instruction for parameter type.
-        auto ParamTy = FTy->getParamType(k);
-        if (ParamTy->isPointerTy()) {
-          auto PointeeTy = ParamTy->getPointerElementType();
-          if (PointeeTy->isStructTy() &&
-              dyn_cast<StructType>(PointeeTy)->isOpaque()) {
-            ParamTy = PointeeTy;
-          }
-        }
 
-        Ops << MkId(lookupType(ParamTy));
-      }
-
-      auto *Inst = new SPIRVInstruction(spv::OpTypeFunction, nextID++, Ops);
-      SPIRVInstList.push_back(Inst);
-      break;
-    }
-    }
+    ID = addSPIRVInst<kTypes>(spv::OpTypeFunction, Ops);
+    break;
   }
-
-  // Generate OpTypeSampledImage.
-  for (auto &ImgTy : getImageTypeList()) {
-    //
-    // Generate OpTypeSampledImage.
-    //
-    // Ops[0] = Image Type ID
-    //
-    SPIRVOperandList Ops;
-
-    Ops << MkId(TypeMap[ImgTy]);
-
-    // Update the image type map.
-    getImageTypeMap()[ImgTy] = nextID;
-
-    auto *Inst = new SPIRVInstruction(spv::OpTypeSampledImage, nextID++, Ops);
-    SPIRVInstList.push_back(Inst);
   }
 
-  // Generate types for pointer-to-local arguments.
-  for (auto spec_id = clspv::FirstLocalSpecId(); spec_id < max_local_spec_id_;
-       ++spec_id) {
-    LocalArgInfo &arg_info = LocalSpecIdInfoMap[spec_id];
-
-    // Generate the spec constant.
-    SPIRVOperandList Ops;
-    Ops << MkId(lookupType(Type::getInt32Ty(Context))) << MkNum(1);
-    SPIRVInstList.push_back(
-        new SPIRVInstruction(spv::OpSpecConstant, arg_info.array_size_id, Ops));
-
-    // Generate the array type.
-    Ops.clear();
-    // The element type must have been created.
-    uint32_t elem_ty_id = lookupType(arg_info.elem_type);
-    assert(elem_ty_id);
-    Ops << MkId(elem_ty_id) << MkId(arg_info.array_size_id);
-
-    SPIRVInstList.push_back(
-        new SPIRVInstruction(spv::OpTypeArray, arg_info.array_type_id, Ops));
+  if (ID) {
+    TypeMap[Ty] = ID;
+  }
+  return ID;
+}
 
-    Ops.clear();
-    Ops << MkNum(spv::StorageClassWorkgroup) << MkId(arg_info.array_type_id);
-    SPIRVInstList.push_back(new SPIRVInstruction(
-        spv::OpTypePointer, arg_info.ptr_array_type_id, Ops));
+void SPIRVProducerPass::GenerateSPIRVTypes(LLVMContext &Context,
+                                           Module &module) {
+  for (Type *Ty : getTypeList()) {
+    getSPIRVType(Ty);
```

Comment:
At some point types can be created on demand but many tests would need updated. So for now still gather a TypeList and generate types up front.